paperless: add package and service

erikarvstedt opened the pull request:
###### Notes for reviewers
[Paperless](https://github.com/danielquinn/paperless) has no setup.py, the main way to use it is to call `manage.py` in a properly setup
Python environment. That's what `${paperless}/bin/paperless` does.

Paperless requires some Python packages that are missing from `nixpkgs`. Currently, they are defined in `extra-python-packages.nix`. Should I add them all to `python-packages.nix`? Do we need an extra git commit for each package?

Here's a sample script for testing:
```bash
dataDir=/tmp/paperless-data
rm -rf $dataDir

nix-build --out-link ./paperless - <<EOF
(import <nixpkgs> {}).paperless.withConfig {
  dataDir = $dataDir;
  config = {
    PAPERLESS_DISABLE_LOGIN = "true";
  };
}
EOF

# Setup DB
./paperless migrate

# Create a test document
$(nix-build --no-out-link '<nixpkgs>' -A imagemagick)/bin/convert \
  -size 400x40 xc:white -font "DejaVu-Sans" -pointsize 20 -fill black -annotate +5+20 \
  "hello world 16-10-2005" $dataDir/consume/doc.png

# Consume document
./paperless document_consumer --oneshot

# Start webinterface
./paperless runserver --noreload localhost:8000

# Cleanup
rm -rf $dataDir ./paperless
```


###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

jfrankenau commented:
Python packages should be globally available. Nowadays each package gets one directory/file. One commit per package would be nice but it may be ok to put them all in one.

Have you thought about creating a matching NixOS module?

dotlambda reviewed (changes requested):
Please create the packages besides paperless itself inside the `pythonPackages` package set. Some of them might already be packaged.

erikarvstedt commented:
Back from vacation, please excuse the delay!

@dotlambda, done.

I've added a service module, but there are a few open issues:
- The web interface (`paperless-server`) runs on the Django dev server. Should I use a proper web server like nginx or Apache?
- By default, the services run with user `nobody`. Should I add a dedicated `paperless` user (via `ids.nix`)?
- In a service preStart script im using `${pkgs.libuuid}/bin/runuser` to run a cmd as the unprivileged user. Is there better, canonical way to do this? `su` is unavailable in the script env.

I'm using [this script](https://gist.github.com/erikarvstedt/d37ba23e336fbcf30f54daafde3085ed) for testing the service.
Run this for a quick demo:
```bash
#! /usr/bin/env bash
set -euo pipefail

nixpkgs=$(nix eval --raw '(builtins.fetchTarball https://github.com/erikarvstedt/nixpkgs/archive/paperless.tar.gz)')

if [[ $PATH != *paperlessDemoEnv* ]]; then
    nix-build --out-link /tmp/paperlessDemoEnv -E '
      with (import <nixpkgs> {});
      let
        extra-container = callPackage (builtins.fetchGit {
          url = "https://github.com/erikarvstedt/extra-container.git";
          rev = "d2d732456d478168d62124ed112396ed28f7d116";
        }) {};
      in
        buildEnv {
          name = "paperlessDemoEnv";
          paths = [ extra-container jq imagemagick ];
        }
    ' > /dev/null
    export PATH=/tmp/paperlessDemoEnv/bin${PATH:+:}$PATH
fi

## Create a container with a Paperless instance.
# For unknown reasons, when using `nixos-container create` the server process
# needs > 30 sec until it opens its socket.
# Work around this by using `extra-container`.
NIX_PATH=nixpkgs=$nixpkgs sudo extra-container create --start <<EOF
{
  containers.pldemo = {
    config = {
      services.paperless = {
        enable = true;
        address = "0.0.0.0";
      };
    };
  };
}
EOF

# Create a test document
sudo convert -size 400x40 xc:white -font "DejaVu-Sans" -pointsize 20 -fill black -annotate +5+20 \
     "hello world 16-10-2005" /var/lib/containers/pldemo/var/lib/paperless/consume/doc.png

# Wait until the document is consumed
while [[ $(curl -s localhost:28981/api/documents/ | jq .count) != 1 ]]; do sleep 0.1; done

# Show documents
curl -s localhost:28981/api/documents/ | jq

sudo extra-container destroy pldemo
```

jfrankenau commented:
@erikarvstedt, I guess for now using Django is OK.

`/var/lib/paperless` should be owned by a real user, so adding a new user to `ids.nix` is necessary.

What about adding another `ExecStart` line to run the unprivileged migration process? Or even extract it to another systemd oneshot service?

jfrankenau commented on nixos/modules/services/misc/paperless.nix:
```diff
@@ -0,0 +1,155 @@
+
+    consumptionDir = mkOption {
+      type = with types; nullOr str;
+      default = null;
```
Set it to the default value mentioned in the description and remove the if-clause below.

erikarvstedt replied:
Good catch! Fixed.

erikarvstedt commented:
The migration has to run in a preStart script so that the server, which is only active when the consumer is active, is disabled during migrations.
I've also thought about a helper oneshot service, but I think that's even more complicated.

But what's the actual benefit of using a dedicated paperless user? Is this documented somewhere in NixOS?

erikarvstedt commented:
Ok, done.

globin commented on pkgs/top-level/python-packages.nix:
```diff
@@ -284,7 +284,9 @@ in {
 
   fire = callPackage ../development/python-modules/fire { };
 
-  globus-sdk = callPackage ../development/python-modules/globus-sdk { };
```
You apparently accidentally deleted this line which causes the CI to fail during evaluation:
https://gist.github.com/GrahamcOfBorg/14241c580774a606e627f780e1809db9

erikarvstedt replied:
Done. Thanks for digging into this, I thought that error came from upstream.

dotlambda commented on pkgs/development/python-modules/django-crispy-forms/default.nix:
```diff
@@ -0,0 +1,19 @@
+    sha256 = "0pv7y648i8iz7mf64gkjizpbx5d01ap2s4vqqa30n38if6wvlljr";
+  };
+
+  doCheck = false;
```
Why? Add a comment. For the other packages as well.

dotlambda commented on pkgs/development/python-modules/langdetect/default.nix:
```diff
@@ -0,0 +1,21 @@
+  version = "1.0.7";
+
+  # Although langdetect is hosted on PyPI, none of the canonical PyPI mirror URLs
+  # (like https://pypi.io/packages/source/l/langdetect/langdetect-1.0.7.tar.gz) are working
```
You have to use `extension = "zip"`.

dotlambda commented on pkgs/development/python-modules/pytest-env/default.nix:
```diff
@@ -0,0 +1,19 @@
+  propagatedBuildInputs = [ pytest ];
+
+  meta = with lib; {
+    description = "pytest plugin used to set environment variables";
```
Capitalize.

dotlambda reviewed (changes requested):
Most of the packages are missing `meta.maintainers`.

FRidh commented on pkgs/development/python-modules/pytest-env/default.nix:
```diff
@@ -0,0 +1,19 @@
+    sha256 = "1hl0ln0cicdid4qjk7mv90lw9xkb0v71dlj7q7rn89vzxxm9b53y";
+  };
+
+  propagatedBuildInputs = [ pytest ];
```
Don't propagate `pytest`, it's a special case. Plugins should only use it for build time of itself.

FRidh commented on pkgs/development/python-modules/python-dotenv/default.nix:
```diff
@@ -0,0 +1,19 @@
+    sha256 = "1mc901wfxd0sxw0baqcb08dh66waarkfcx4r009ic4apa8c3d5sh";
+  };
+
+  doCheck = false;
```
Why are the tests disabled? Please include a comment.

FRidh commented on pkgs/development/python-modules/filemagic/default.nix:
```diff
@@ -0,0 +1,24 @@
+      "'${file}/lib/libmagic${stdenv.hostPlatform.extensions.sharedLibrary}'"
+  '';
+
+  doCheck = false;
```
Why are the tests disabled? Please include a comment.

FRidh commented on pkgs/development/python-modules/django-flat-responsive/default.nix:
```diff
@@ -0,0 +1,19 @@
+    sha256 = "0x3439m2bim8r0xldx99ry0fksfyv39k8bffnwpvahf500ksl725";
+  };
+
+  doCheck = false;
```
Why are the tests disabled? Please include a comment.

FRidh commented on pkgs/development/python-modules/django-filter/default.nix:
```diff
@@ -0,0 +1,19 @@
+    sha256 = "0slpfqfhnjrzlrb6vmswyhrzn01p84s16j2x1xib35gg4fxg23pc";
+  };
+
+  doCheck = false;
```
Why are the tests disabled? Please include a comment.

FRidh commented on pkgs/applications/office/paperless/default.nix:
```diff
@@ -0,0 +1,204 @@
+        (source $out/share/paperless/setup-env.sh
+         export HOME=$(pwd)
+         cd $paperlessSrc
+         pytest -p no:cacheprovider)
```
Why without cache? Explain with a comment.

erikarvstedt commented:
Thanks for your great feedback!

All the remaining python pkgs tests are now enabled.

What's the right way to run the [tests for django-crispy-forms](https://github.com/NixOS/nixpkgs/blob/8e283dce84fdf8d103107977150741fd5663d097/pkgs/development/python-modules/django-crispy-forms/default.nix#L15)?
I have to make sure not to import sources from both `$PWD` and `$out`, otherwise I get a `py._path.local.LocalPath.ImportMismatchError`.
Another way would be
```
DJANGO_SETTINGS_MODULE=crispy_forms.tests.test_settings pytest $out/lib/python*/site-packages/crispy_forms/tests
```
but that would generate needless `.pytest_cache` files in `$out`.

erikarvstedt commented:
Thanks for your great feedback!

All the remaining python pkg tests are now enabled.

What's the right way to run the [tests for django-crispy-forms]()?
We have to make sure not to import sources from both `$PWD` and `$out`, otherwise we get a `py._path.local.LocalPath.ImportMismatchError`.
Another way would be
```
DJANGO_SETTINGS_MODULE=crispy_forms.tests.test_settings pytest $out/lib/python*/site-packages/crispy_forms/tests
```
but that would generate needless `.pytest_cache` files in `$out`.

erikarvstedt commented:
I've now squashed all the fixup commits.
[See here for a history](https://github.com/erikarvstedt/nixpkgs/commits/paperless-history) of all changes since your last review.

erikarvstedt commented:
I've addressed some remaining quirks and corner cases.
This PR is now finished. Let me know if you spot anything left to be done.

(I'll resolve the conflict in `ids.nix` when the unstable channel has caught up to
the latest changes.)

erikarvstedt commented:
We're currently blocked by https://hydra.nixos.org/build/84873905

GrahamcOfBorg commented:
<!--REQUEST_ID=74468927-9e27-464a-a8f6-7768661c3c1a-->
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44086&attempt_id=029183f0-f659-42c8-9bfc-77846a7a7d35)

Attempted: paperless

<details><summary>Partial log (click to expand)</summary><p>

```
SKIP [1] tests/tests_libtesseract.py:212: unconditional skip
SKIP [1] tests/tests_tesseract.py:66: unconditional skip
SKIP [1] tests/tests_tesseract.py:69: unconditional skip
SKIP [1] tests/tests_tesseract.py:188: unconditional skip

==================== 53 passed, 15 skipped in 30.40 seconds ====================
pytestcachePhase
cannot build derivation '/nix/store/n9mx6kjx675fdrz254a0r59dsdm55jms-python3-3.7.1-env.drv': 4 dependencies couldn't be built
cannot build derivation '/nix/store/38x4s5z4myhr5c1k7q2i5qf29vmfyjqf-paperless-2.6.0-pre.drv': 1 dependencies couldn't be built
error: build of '/nix/store/38x4s5z4myhr5c1k7q2i5qf29vmfyjqf-paperless-2.6.0-pre.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=74468927-9e27-464a-a8f6-7768661c3c1a-->
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44086&attempt_id=79a58d7e-dd94-4a5a-a124-fe68fddfd58e)

Attempted: paperless

<details><summary>Partial log (click to expand)</summary><p>

```
=============================== warnings summary ===============================
tests/tests_cuneiform.py::TestContext::()::test_version
  source:36: DeprecationWarning: invalid escape sequence \w

-- Docs: https://docs.pytest.org/en/latest/warnings.html
========= 2 failed, 51 passed, 15 skipped, 1 warnings in 50.52 seconds =========
builder for '/nix/store/skzvgnaq0yl16aws9j23h1hz9sqigmz8-python3.7-pyocr-0.5.3.drv' failed with exit code 1
cannot build derivation '/nix/store/0h7i7n5b43gadg9wnphp65f78x0lpm6q-python3-3.7.1-env.drv': 5 dependencies couldn't be built
cannot build derivation '/nix/store/p55jyy54v7rkmnm3kd6jar5m45va93jp-paperless-2.6.0-pre.drv': 1 dependencies couldn't be built
error: build of '/nix/store/p55jyy54v7rkmnm3kd6jar5m45va93jp-paperless-2.6.0-pre.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=c71ffcc6-78ef-421a-82c4-dbf68a39dd16-->
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44086&attempt_id=611ef132-f1fc-4d03-b4bc-b304f7e770dc)

Attempted: paperless

<details><summary>Partial log (click to expand)</summary><p>

```

tests/tests_cuneiform.py::TestContext::()::test_version
  source:36: DeprecationWarning: invalid escape sequence \w

-- Docs: https://docs.pytest.org/en/latest/warnings.html
========= 2 failed, 51 passed, 15 skipped, 2 warnings in 54.56 seconds =========
builder for '/nix/store/lbq7q8bdpygmgwv6jkrc6xi9gb5gxn2k-python3.6-pyocr-0.5.3.drv' failed with exit code 1
cannot build derivation '/nix/store/2ahpbiaci0a3qbrpmywa96fz69830nn1-python3-3.6.7-env.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/dg0mf8in8y8ak4xw0881lq8a7295lpx9-paperless-2.6.0-pre.drv': 1 dependencies couldn't be built
error: build of '/nix/store/dg0mf8in8y8ak4xw0881lq8a7295lpx9-paperless-2.6.0-pre.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=94b2194d-8b26-4b0a-8b5c-ce7f5d2b73f9-->
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44086&attempt_id=cdbea77e-bc18-4aba-92d7-4b2fd6e2965a)

Attempted: paperless

<details><summary>Partial log (click to expand)</summary><p>

```
writing manifest file 'python_dotenv.egg-info/SOURCES.txt'
running build_ext

----------------------------------------------------------------------
Ran 0 tests in 0.000s

OK
cannot build derivation '/nix/store/ryga3y4zv58brbnyr4sqfsp84yjw7zdd-python3-3.6.7-env.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/g5qi0smwz5jspybngpp56ai31j2mwmxf-paperless-2.6.0-pre.drv': 1 dependencies couldn't be built
error: build of '/nix/store/g5qi0smwz5jspybngpp56ai31j2mwmxf-paperless-2.6.0-pre.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=fea2d529-fdb8-419f-bd2f-234c606b9103-->
No attempt on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44086&attempt_id=09d4acf9-573a-4508-9c97-0bbb8a1e4d93)

The following builds were skipped because they don't evaluate on x86_64-linux: tests.paperless

<details><summary>Partial log (click to expand)</summary><p>

```
error: while evaluating 'recursiveUpdate' at /var/lib/gc-of-borg/nix-root/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/gleber-bastion/lib/attrsets.nix:415:26, called from /var/lib/gc-of-borg/nix-root/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/gleber-bastion/lib/attrsets.nix:148:28:
while evaluating 'recursiveUpdateUntil' at /var/lib/gc-of-borg/nix-root/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/gleber-bastion/lib/attrsets.nix:384:37, called from /var/lib/gc-of-borg/nix-root/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/gleber-bastion/lib/attrsets.nix:416:5:
while evaluating 'zipAttrsWith' at /var/lib/gc-of-borg/nix-root/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/gleber-bastion/lib/attrsets.nix:347:21, called from /var/lib/gc-of-borg/nix-root/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/gleber-bastion/lib/attrsets.nix:394:8:
while evaluating 'zipAttrsWithNames' at /var/lib/gc-of-borg/nix-root/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/gleber-bastion/lib/attrsets.nix:332:33, called from /var/lib/gc-of-borg/nix-root/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/gleber-bastion/lib/attrsets.nix:347:27:
while evaluating anonymous function at /var/lib/gc-of-borg/nix-root/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/gleber-bastion/lib/lists.nix:113:41, called from /var/lib/gc-of-borg/nix-root/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/gleber-bastion/lib/attrsets.nix:347:46:
while evaluating the attribute 'paperless' at /var/lib/gc-of-borg/nix-root/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/gleber-bastion/nixos/tests/all-tests.nix:161:3:
while evaluating 'handleTest' at /var/lib/gc-of-borg/nix-root/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/gleber-bastion/nixos/tests/all-tests.nix:17:22, called from /var/lib/gc-of-borg/nix-root/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/gleber-bastion/nixos/tests/all-tests.nix:161:15:
while evaluating 'discoverTests' at /var/lib/gc-of-borg/nix-root/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/gleber-bastion/nixos/tests/all-tests.nix:13:19, called from /var/lib/gc-of-borg/nix-root/repo/38dca4e3aa6bca43ea96d2fcc04e8229/builder/gleber-bastion/nixos/tests/all-tests.nix:18:5:
access to path '/nix/store/6mqim37559zn5v944gy6yi91k4zifwvv-gleber-bastion' is forbidden in restricted mode

```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=59fa9032-5c9b-46b6-aa53-36df877dbd09-->
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44086&attempt_id=7e92b20d-89ec-4d6a-8f49-a43194a4c539)

Attempted: paperless

<details><summary>Partial log (click to expand)</summary><p>

```
Couldn't find index page for 'mock' (maybe misspelled?)
Scanning index of all packages (this may take a while)
Reading https://pypi.org/simple/
Download error on https://pypi.org/simple/: [Errno -2] Name or service not known -- Some packages may not be found!
No local packages or working download links found for mock
error: Could not find suitable distribution for Requirement.parse('mock')
builder for '/nix/store/xf0px8saq1vs6vj2r4hrlmqcm2bk4kaa-python3.6-filemagic-1.6.drv' failed with exit code 1
cannot build derivation '/nix/store/ryga3y4zv58brbnyr4sqfsp84yjw7zdd-python3-3.6.7-env.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/g5qi0smwz5jspybngpp56ai31j2mwmxf-paperless-2.6.0-pre.drv': 1 dependencies couldn't be built
error: build of '/nix/store/g5qi0smwz5jspybngpp56ai31j2mwmxf-paperless-2.6.0-pre.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=59fa9032-5c9b-46b6-aa53-36df877dbd09-->
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44086&attempt_id=503b0e61-83f8-46a6-9b3e-9952d24ffaff)

Attempted: paperless

<details><summary>Partial log (click to expand)</summary><p>

```

tests/tests_cuneiform.py::TestContext::()::test_version
  source:36: DeprecationWarning: invalid escape sequence \w

-- Docs: https://docs.pytest.org/en/latest/warnings.html
========= 2 failed, 51 passed, 15 skipped, 2 warnings in 50.10 seconds =========
builder for '/nix/store/lbq7q8bdpygmgwv6jkrc6xi9gb5gxn2k-python3.6-pyocr-0.5.3.drv' failed with exit code 1
cannot build derivation '/nix/store/2ahpbiaci0a3qbrpmywa96fz69830nn1-python3-3.6.7-env.drv': 2 dependencies couldn't be built
cannot build derivation '/nix/store/dg0mf8in8y8ak4xw0881lq8a7295lpx9-paperless-2.6.0-pre.drv': 1 dependencies couldn't be built
error: build of '/nix/store/dg0mf8in8y8ak4xw0881lq8a7295lpx9-paperless-2.6.0-pre.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=c71ffcc6-78ef-421a-82c4-dbf68a39dd16-->
Failure on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44086&attempt_id=5b148f01-8fc3-417c-9a2f-023125b407d6)

Attempted: paperless

<details><summary>Partial log (click to expand)</summary><p>

```
tests/management/commands/shell_plus_tests/test_collision_resolver.py::CRTestCase::test_installed_bad_order_collision_resolver
  /build/source/tests/management/commands/shell_plus_tests/test_collision_resolver.py:160: DeprecationWarning: Please use assertRaisesRegex instead.
    with self.assertRaisesRegexp(AssertionError, "You must define APP_PRIORITIES in your resolver class!"):

-- Docs: https://docs.pytest.org/en/latest/warnings.html
============= 237 passed, 9 skipped, 39 warnings in 33.65 seconds ==============
pytestcachePhase
cannot build derivation '/nix/store/ryga3y4zv58brbnyr4sqfsp84yjw7zdd-python3-3.6.7-env.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/g5qi0smwz5jspybngpp56ai31j2mwmxf-paperless-2.6.0-pre.drv': 1 dependencies couldn't be built
error: build of '/nix/store/g5qi0smwz5jspybngpp56ai31j2mwmxf-paperless-2.6.0-pre.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=a4a5beaa-208b-4d76-883e-4d992ffc20aa-->
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44086&attempt_id=4639b820-dc6e-41e5-8337-749f94582597)

Attempted: paperless

<details><summary>Partial log (click to expand)</summary><p>

```

tests/tests_cuneiform.py::TestContext::()::test_version
  source:36: DeprecationWarning: invalid escape sequence \w

-- Docs: https://docs.pytest.org/en/latest/warnings.html
========= 2 failed, 51 passed, 15 skipped, 2 warnings in 69.41 seconds =========
builder for '/nix/store/lbq7q8bdpygmgwv6jkrc6xi9gb5gxn2k-python3.6-pyocr-0.5.3.drv' failed with exit code 1
cannot build derivation '/nix/store/f8snvsf5lljadjj2j80lwskh3wppvhqg-python3-3.6.7-env.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/1m4iskz0q3wav0j625kvdqdiznkx8nqn-paperless-2.6.0-pre.drv': 1 dependencies couldn't be built
error: build of '/nix/store/1m4iskz0q3wav0j625kvdqdiznkx8nqn-paperless-2.6.0-pre.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=a4a5beaa-208b-4d76-883e-4d992ffc20aa-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44086&attempt_id=82ca79d9-bc6e-4bfc-b7d9-6a4cb71e1c2e)

Attempted: paperless

<details><summary>Partial log (click to expand)</summary><p>

```
documents/tests/test_mail.py ...                                         [ 47%]
documents/tests/test_matchables.py ..........                            [ 59%]
documents/tests/test_models.py ..                                        [ 61%]
paperless_tesseract/tests/test_date.py ............................      [ 95%]
paperless_tesseract/tests/test_ocr.py ..                                 [ 97%]
paperless_tesseract/tests/test_signals.py ..                             [100%]

==================== 82 passed, 2 skipped in 76.62 seconds =====================
cannot link '/nix/store/.links/1iaj3s31gnwicm04pmsxkns696qi1dfllkmgnxn1fw74c8a5ns7f' to '/nix/store/7r8lwg4p9yj9y3wklw80h29gmjdlkr3x-paperless-2.6.0-pre/share/paperless/manage.py': No space left on device
/nix/store/7r8lwg4p9yj9y3wklw80h29gmjdlkr3x-paperless-2.6.0-pre
```
</p></details>

erikarvstedt commented:
Is there anything left to do to get this merged?

dotlambda commented:
@GrahamcOfBorg build paperless

GrahamcOfBorg commented:
<!--REQUEST_ID=c1583e4c-c216-4a1b-a15b-1e920f5e21f4-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44086&attempt_id=5cdaf50c-6903-411f-af3c-da392f2e4aa5)

Attempted: paperless

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/7r8lwg4p9yj9y3wklw80h29gmjdlkr3x-paperless-2.6.0-pre
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=c1583e4c-c216-4a1b-a15b-1e920f5e21f4-->
No attempt on x86_64-darwin [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44086&attempt_id=f9482f41-2aab-42fe-a92e-fcbff8d508c5)

The following builds were skipped because they don't evaluate on x86_64-darwin: paperless

<details><summary>Partial log (click to expand)</summary><p>

```

a) For `nixos-rebuild` you can set
  { nixpkgs.config.allowUnsupportedSystem = true; }
in configuration.nix to override this.

b) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
  { allowUnsupportedSystem = true; }
to ~/.config/nixpkgs/config.nix.


```
</p></details>

dotlambda commented on pkgs/applications/office/paperless/default.nix:
```diff
@@ -0,0 +1,222 @@
+
+      ### Compile
+
+      ${python}/bin/python -m compileall $srcDir
```
```suggestion
      ${python.interpreter} -m compileall $srcDir
```
The same multiple times.

erikarvstedt replied:
Done.

dotlambda commented on pkgs/applications/office/paperless/default.nix:
```diff
@@ -0,0 +1,222 @@
+      termcolor
+    ] ++ (optional stdenv.isLinux inotify_simple));
+
+    buildCommand = let
```
Why not `buildPhase`?

erikarvstedt replied:
Done, see commit `use phases instead of buildCommand`

dotlambda commented on pkgs/applications/office/paperless/default.nix:
```diff
@@ -0,0 +1,222 @@
+, tesseract
+, unpaper
+
+, enableTests ? true
```
```suggestion
```

erikarvstedt replied:
Done.

dotlambda commented on pkgs/applications/office/paperless/default.nix:
```diff
@@ -0,0 +1,222 @@
+      export paperlessSrc=$out/share/paperless
+      EOF
+
+      ${optionalString enableTests ''
```
The following lines belong into `checkPhase`.

erikarvstedt replied:
Done.

dotlambda commented on pkgs/development/python-modules/django-filter/default.nix:
```diff
@@ -0,0 +1,22 @@
+
+  checkInputs = [ django django-crispy-forms djangorestframework mock pytz ];
+
+  checkPhase = "python runtests.py";
```
```suggestion
  checkPhase = "${python.interpreter} runtests.py";
```

erikarvstedt replied:
Done. But isn't this equivalent, as the same `python` is already a input to the derivation?

dotlambda commented on nixos/modules/services/misc/paperless.nix:
```diff
@@ -0,0 +1,181 @@
+
+    manage = mkOption {
+      type = types.package;
+      readOnly = true;
```
Why do you provide the option?

erikarvstedt replied:
To aid discoverability and to ease scripting: The manage script is also symlinked to dataDir, but it's only available when the service has been run at least once – which might be confusing. But I'm fine with removing it if you'd recommend it.
By the way: `autoSetupDB` may also be a fine candidate for removal, as there are few good reasons to disable it.

Thanks a lot for having another look at this PR, I really appreciate your attention to the details!

GrahamcOfBorg commented:
<!--REQUEST_ID=c1583e4c-c216-4a1b-a15b-1e920f5e21f4-->
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44086&attempt_id=7a007a77-d3e6-4338-8d4f-ec972426f876)

Attempted: paperless

<details><summary>Partial log (click to expand)</summary><p>

```

tests/tests_cuneiform.py::TestContext::()::test_version
  source:36: DeprecationWarning: invalid escape sequence \w

-- Docs: https://docs.pytest.org/en/latest/warnings.html
========= 2 failed, 51 passed, 15 skipped, 2 warnings in 55.46 seconds =========
builder for '/nix/store/lbq7q8bdpygmgwv6jkrc6xi9gb5gxn2k-python3.6-pyocr-0.5.3.drv' failed with exit code 1
cannot build derivation '/nix/store/f8snvsf5lljadjj2j80lwskh3wppvhqg-python3-3.6.7-env.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/1m4iskz0q3wav0j625kvdqdiznkx8nqn-paperless-2.6.0-pre.drv': 1 dependencies couldn't be built
error: build of '/nix/store/1m4iskz0q3wav0j625kvdqdiznkx8nqn-paperless-2.6.0-pre.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=608e9129-48ef-4176-9852-9e1f9ff506dd-->
Failure on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44086&attempt_id=cb38b63e-1606-4080-bb8c-36eaeb667370)

Attempted: paperless

<details><summary>Partial log (click to expand)</summary><p>

```

tests/tests_cuneiform.py::TestContext::()::test_version
  source:36: DeprecationWarning: invalid escape sequence \w

-- Docs: https://docs.pytest.org/en/latest/warnings.html
========= 2 failed, 51 passed, 15 skipped, 2 warnings in 57.65 seconds =========
builder for '/nix/store/lbq7q8bdpygmgwv6jkrc6xi9gb5gxn2k-python3.6-pyocr-0.5.3.drv' failed with exit code 1
cannot build derivation '/nix/store/0i2lq6ff2v4nsqrv9lacrwyav6g88s0p-python3-3.6.7-env.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/dmvaakpldmc4x2bydjrfny7b20pyff9c-paperless-2.6.0-pre.drv': 1 dependencies couldn't be built
error: build of '/nix/store/dmvaakpldmc4x2bydjrfny7b20pyff9c-paperless-2.6.0-pre.drv' failed
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=608e9129-48ef-4176-9852-9e1f9ff506dd-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44086&attempt_id=5c4d1783-c8eb-4b14-a07d-52782ad4f9a2)

Attempted: paperless

<details><summary>Partial log (click to expand)</summary><p>

```
documents/tests/test_mail.py ...                                         [ 47%]
documents/tests/test_matchables.py ..........                            [ 59%]
documents/tests/test_models.py ..                                        [ 61%]
paperless_tesseract/tests/test_date.py ............................      [ 95%]
paperless_tesseract/tests/test_ocr.py ..                                 [ 97%]
paperless_tesseract/tests/test_signals.py ..                             [100%]

==================== 82 passed, 2 skipped in 77.93 seconds =====================
cannot link '/nix/store/.links/1iaj3s31gnwicm04pmsxkns696qi1dfllkmgnxn1fw74c8a5ns7f' to '/nix/store/0mkh0k9f7vy09hwcmcnzc9gs9n6ydzjh-paperless-2.6.0-pre/share/paperless/manage.py': No space left on device
/nix/store/0mkh0k9f7vy09hwcmcnzc9gs9n6ydzjh-paperless-2.6.0-pre
```
</p></details>

erikarvstedt commented:
@dotlambda, should I rebase to resolve the merge conflict and get grahamcofborg building again or do we finish the review first?

erikarvstedt commented:
Rebased to `master`.

dotlambda commented on pkgs/applications/office/paperless/default.nix:
```diff
@@ -0,0 +1,222 @@
+       inherit setupEnv;
+     };
+in
+  paperless
```
Please put the `mkDerivation` here, not inside the `let` statement.

dotlambda commented on pkgs/applications/office/paperless/default.nix:
```diff
@@ -0,0 +1,222 @@
+      pytest
+      pytest-django
+      pytest-env
+      # pytest_xdist # Disabled, broken in nixpkgs
```
I think this is no longer the case.

erikarvstedt replied:
It's still broken: https://github.com/NixOS/nixpkgs/issues/45060#issuecomment-448397355
Could you reopen this issue?

dotlambda replied:
Now it's finally fixed.

dotlambda commented on pkgs/applications/office/paperless/default.nix:
```diff
@@ -0,0 +1,222 @@
+          mkdir -p "$PAPERLESS_CONSUMPTION_DIR" \
+                   "$PAPERLESS_MEDIADIR" \
+                   "$PAPERLESS_STATICDIR" \
+                   "$PAPERLESS_DBDIR"
```
This means the directories are created every time we run paperless? Wouldn't it be better to move all this stuff (everything from `withConfig`) into the service? Setting something up should not be done by the package if it's not what upstream does.

erikarvstedt replied:
Done.

dotlambda commented on pkgs/applications/office/paperless/default.nix:
```diff
@@ -0,0 +1,222 @@
+          doInstallCheck = false;
+        });
+
+  withConfig = { config ? {}, dataDir ? null, paperlessPkg ? paperless, extraCmds ? "" }:
```
Maybe there's a better name for `config`?

dotlambda commented on pkgs/applications/office/paperless/default.nix:
```diff
@@ -0,0 +1,222 @@
+
+with lib;
+let
+  paperless = stdenv.mkDerivation rec {
```
Can't we use `buildPythonApplication` here? This would allow us to specify `checkInputs` etc.
Of course, you'd have to override most of the phases.
I don't like that stuff like `pytest` is in the `pythonEnv` and hence not only available during the tests.

erikarvstedt replied:
I've separated the runtime and check inputs. 
The current use of `stdenv.mkDerivation` seems simpler to me than using `buildPythonApplication` (with `format = "other"`), so I haven't changed that.

dotlambda commented on pkgs/development/python-modules/pytest-env/default.nix:
```diff
@@ -0,0 +1,19 @@
+    sha256 = "1hl0ln0cicdid4qjk7mv90lw9xkb0v71dlj7q7rn89vzxxm9b53y";
+  };
+
+  nativeBuildInputs = [ pytest ];
```
```suggestion
   checkInputs = [ pytest ];
```

erikarvstedt replied:
Fixed.

dotlambda commented on pkgs/applications/office/paperless/default.nix:
```diff
@@ -0,0 +1,222 @@
+
+      # Skip failing test (see Paperless pull request 450)
+      sed -i '1s/^/import pytest\n/; /def test_get_text_3_png/i \
+          @pytest.mark.skip' $srcDir/paperless_tesseract/tests/test_date.py
```
This stuff belongs into `postPatch`.

erikarvstedt replied:
[Fixed.](https://github.com/NixOS/nixpkgs/pull/44086/commits/e9d956afd95496951ac92f4c66ad46cec1a2a966)

dotlambda replied:
You will really need to use `postPatch` if my suggestion above (applying the patch using `fetchpatch`) is followed.

dotlambda reviewed (commented):
I guess the service is okay but the package just seems more complicated than a package ought to be. Please move that stuff into the service.
Also, please add `meta.maintainers` to the new python packages.

PS: Sorry it's taking me so long to review this.

erikarvstedt commented:
My rationale for providing `paperless.withConfig` in `nixpkgs` is the following:
Paperless is barely usable in it's raw state. It's primarily meant to be run in a Docker container and the included Dockerfile contains all the setup helpers that I've recreated in `withConfig`. (The Arch AUR package has a similar approach.)
Due to Paperless' complex dependencies and the impure nature of Docker, we get occasional bug reports from self-building Docker users that are related to subtle package incompatibilities. These are mostly non-reproducible and only solved by a rebuild of the container.
Nix could truly shine here and a convenient `pkgs.paperless` would be a great way to introduce new folks to nix.

I agree that the config helper should be separated from the core package. Would it be okay if I'd put it in a `wrapper.nix`?

> Sorry it's taking me so long to review this.

No problem. I understand that these long PRs have to take a back seat sometimes.
Thanks for all your contributions to nixpkgs!

dotlambda commented:
> I agree that the config helper should be separated from the core package. Would it be okay if I'd put it in a wrapper.nix?

That would probably be acceptable. If Arch is doing the same, it seems reasonable.

FRidh commented:
A point of worry. According to https://github.com/danielquinn/paperless#project-status the project is still maintained but no longer actively developed. That's fine, however, I would like to know whether and how often it is updated to handle newer versions of dependencies. The reason I would like to know this is because we update Python packages continuously and don't keep older versions around.

erikarvstedt commented:
@FRidh, as you noted, Paperless is considered feature-complete but is still maintained to work with current versions of its dependencies. E.g., Paperless switched to Django 2.0 just recently.

Besides that, I'm dedicated to provide NixOS support for the foreseeable future. That means updating Paperless, or, if that's infeasible for whatever reason, adding private versioned module pkgs like in `home-assistant`.

@dotlambda, regarding the service: To make things simpler, I'd like to remove the option `autoSetupDB`, as there's hardly a good reason to disable it.
To justify the `manage` option, I wrote:

> To aid discoverability and to ease scripting: The manage script is also symlinked to dataDir,
> but it's only available when the service has been run at least once – which might be confusing.
> But I'm fine with removing it if you'd recommend it.

What do you think?

dotlambda commented on pkgs/applications/office/paperless/default.nix:
```diff
@@ -0,0 +1,170 @@
+#   source ${paperless}/share/paperless/setup-env.sh
+#   PYTHONPATH=$paperlessSrc gunicorn paperless.wsgi
+
+with lib;
```
I don't think we need this in more than one place. Please get rid of `with` and use `lib` explicily.

dotlambda reviewed (changes requested):
I'm confident this will be my last review.
However, please don't forget to squash your additional commits into the appropriate ones.

dotlambda commented:
You're the expert. If you deem `autoSetupDB` useless then just remove it. However, this will also mean that this feature is no longer documented. You may want to document it somewhere else.
If the `manage` option is useful to you feel free to keep it.

dotlambda commented on pkgs/applications/office/paperless/default.nix:
```diff
@@ -0,0 +1,170 @@
+      # Remove pytest command arguments for pytest_xdist, which is broken in nixpkgs
+      sed -i 's/ -n auto//' $srcDir/setup.cfg
+
+      # Skip failing test (see Paperless pull request 450)
```
Sorry, I did not really look at the patch. Seems like it won't have a meaningful effect on our side. Instead you should update the comment to state that the disabled tests requires German language support for Tesseract.

erikarvstedt replied:
The German lang support is irrelevant to the test failure.
The patch that fixes the test failure consists of a large `GIT binary patch` that can't be applied with `fetchpatch`.
Because the patch will be included in the next Paperless release, I chose to just skip the test for now.
I've moved the relevant code to `postPatch`: https://github.com/NixOS/nixpkgs/pull/44086/commits/55aa8bda951841b7c87c31cfedb4e2b4a6fe4561

erikarvstedt commented:
Rebased to `master`.

erikarvstedt commented on pkgs/applications/office/paperless/default.nix:
```diff
@@ -0,0 +1,170 @@
+        # The user has provided a custom tesseract derivation that might be
+        # missing some languages that are required for PyOCR's tests. Disable them to
+        # avoid build errors.
+        pyocr.overrideAttrs (attrs: {
```
Fixed in https://github.com/NixOS/nixpkgs/pull/44086/commits/a9b3b217d010d78d67a476be642f48daa7547b0d

erikarvstedt commented on pkgs/applications/office/paperless/default.nix:
```diff
@@ -0,0 +1,170 @@
+        # missing some languages that are required for PyOCR's tests. Disable them to
+        # avoid build errors.
+        pyocr.overrideAttrs (attrs: {
+          doInstallCheck = false;
```
Fixed in https://github.com/NixOS/nixpkgs/pull/44086/commits/a9b3b217d010d78d67a476be642f48daa7547b0d

erikarvstedt commented on pkgs/applications/office/paperless/default.nix:
```diff
@@ -0,0 +1,170 @@
+
+    inherit python;
+
+    runtimePackages = with python.pkgs; [
```
I've put these in on purpose. The drv attrs can be very useful for customization (by overriding them or reading their values) and even for developing.
Example: Run tests while hacking Paperless with 
```
${pkgs.paperless.pythonCheckEnv}/bin/pytest /my/dev-paperless
```

I'd vote for keeping `python`, `pythonEnv`, `pythonCheckEnv`. If you want, I can remove `{runtime,check}Packages`.

erikarvstedt commented on pkgs/applications/office/paperless/withConfig.nix:
```diff
@@ -0,0 +1,58 @@
+{ paperless, stdenv, lib, writeScript}:
```
Fixed by using writeBash: https://github.com/NixOS/nixpkgs/pull/44086/commits/ba8843d2c9fb6af7eeaa376c158f8b54bff2e70f

erikarvstedt commented on pkgs/applications/office/paperless/withConfig.nix:
```diff
@@ -0,0 +1,58 @@
+{ config ? {}, dataDir ? null, paperlessPkg ? paperless, extraCmds ? "" }:
+with lib;
+let
+  dir = toString dataDir;
```
So that `dataDir` can also be a path. Should I remove it?

dotlambda replied:
Yes, I don't think we usually do this.

erikarvstedt commented on pkgs/applications/office/paperless/withConfig.nix:
```diff
@@ -0,0 +1,58 @@
+    PAPERLESS_DBDIR = "${dir}";
+  }) // config;
+
+  envVarDefs = mapAttrsToList (n: v: ''export ${n}="${toString v}"'') envVars;
```
This was intentional. See the description of `extraConfig` where the option value is a shell statement:
```
extraConfig.PAPERLESS_PASSPHRASE = "$(< /etc/my_passphrase_file)";
```
`PAPERLESS_PASSPHRASE` and `PAPERLESS_EMAIL_SECRET` are the only options where the
shell evaluation would be relevant.
Curiously, these are also the only options where their literal value could collide with the shell syntax.
Both options are rather obscure and don't warrant dedicated nixos options, so I've left the code unchanged and added a note regarding the shell evaluation in the description: https://github.com/NixOS/nixpkgs/pull/44086/commits/554e5f75d9fd79543428159bd2f1738d35989f3c

erikarvstedt commented on pkgs/applications/office/paperless/withConfig.nix:
```diff
@@ -0,0 +1,58 @@
+    ''}
+  '';
+
+  runPaperless = writeScript "paperless" ''
```
Fixed in https://github.com/NixOS/nixpkgs/pull/44086/commits/ba8843d2c9fb6af7eeaa376c158f8b54bff2e70f

dotlambda commented on nixos/modules/services/misc/paperless.nix:
```diff
@@ -0,0 +1,194 @@
+      '';
+
+      # Auto-migrate on first run or if the package has changed
+      setupDB = pkgs.writeScript "setupDB" ''
```
```suggestion
      setupDB = pkgs.writeBash "setupDB" ''
```

erikarvstedt replied:
`writeBash` would imply an additional `set -e` line in the script. I don't see the gain here.

erikarvstedt replied:
Sorry, that was a misfire. Of course, `writeBash` is a better fit. Fixed: https://github.com/NixOS/nixpkgs/pull/44086/commits/c4209b00e36d3e84df108b237f26c46af50acad7

dotlambda reviewed (approved):
Sorry this took me so long.

LGTM. However, I'd prefer to have someone else have a look at this. @worldofpeace?

Also, you need to rebase on master.

worldofpeace commented:
@dotlambda Anything in particular that I should check?

If it's generally everything I'd say having this squashed back together would make it easier so I could review commit by commit.

worldofpeace commented on pkgs/development/python-modules/fuzzywuzzy/default.nix:
```diff
@@ -0,0 +1,23 @@
+{ lib, buildPythonPackage, fetchPypi, python-Levenshtein
```
I've seen this in a couple prs now so I'm just going to push it to master.

worldofpeace replied:
Pushed in 4094d4c5b1cf1df9af20f29aa26082c5a9d03b66

worldofpeace commented on pkgs/development/python-modules/fuzzywuzzy/default.nix:
```diff
@@ -0,0 +1,23 @@
+
+  propagatedBuildInputs = [ python-Levenshtein ];
+
+  checkInputs = [ pytest hypothesis enum34 pycodestyle ];
```
I'll be pushing this without `enum34` because it's only needed for the tests in python 3.3.

worldofpeace commented on pkgs/development/python-modules/filemagic/default.nix:
```diff
@@ -0,0 +1,29 @@
+      "'${file}/lib/libmagic${stdenv.hostPlatform.extensions.sharedLibrary}'"
+  '';
+
+  checkInputs = [ mock ];
```
We need to constrain `unittest2` like https://github.com/aliles/filemagic/blob/138649062f769fb10c256e454a3e94ecfbf3017b/setup.py#L54

erikarvstedt replied:
Fixed: https://github.com/NixOS/nixpkgs/commit/78af4fb103a1e3748ef86d6f8c174ec11cb64f9f

worldofpeace commented on pkgs/development/python-modules/inotify_simple/default.nix:
```diff
@@ -0,0 +1,21 @@
+{ lib, buildPythonPackage, fetchPypi }:
+
+buildPythonPackage rec {
+  pname = "inotify_simple";
```
I don't like this being `inotify_simple`.

```suggestion
  pname = "inotify-simple";
```

erikarvstedt replied:
I also prefer dashes, but I suggest we should keep the original package name, as it improves discoverability and searchability. (That's my strong preference after all my experiences with searching for nixpgs.)

worldofpeace replied:
I'd really like to not have any additional packages named like this.

And it's not completely a preference as we have a guideline of sorts for this.
> https://nixos.org/nixpkgs/manual/#contributing
> Attribute names in python-packages.nix should be normalized according to PEP 0503. This means that characters should be converted to lowercase and . and _ should be replaced by a single - (foo-bar-baz instead of Foo__Bar.baz )


So I'd think anyone searching for a python package in nixpkgs should expect their names to be normalized in this manner.

erikarvstedt replied:
Sorry, I missed that contrib guideline.
Fixed: https://github.com/NixOS/nixpkgs/pull/44086/commits/93d0511927c88fd4816832129532c3c4040959ce

worldofpeace commented on pkgs/top-level/python-packages.nix:
```diff
@@ -410,6 +414,8 @@ in {
 
   imutils = callPackage ../development/python-modules/imutils { };
 
+  inotify_simple = callPackage ../development/python-modules/inotify_simple { };
```
```suggestion
  inotify-simple = callPackage ../development/python-modules/inotify-simple { };
```

Please be sure to change the directory name also.

worldofpeace commented on pkgs/development/python-modules/python-dotenv/default.nix:
```diff
@@ -0,0 +1,20 @@
+
+  meta = with lib; {
+    description = "Add .env support to your django/flask apps in development and deployments";
+    homepage = http://github.com/theskumar/python-dotenv;
```
```suggestion
    homepage = https://github.com/theskumar/python-dotenv;
```

erikarvstedt replied:
Fixed: https://github.com/NixOS/nixpkgs/commit/67acfe6fc4e41da511cb6e1bd9347189cf9247dd

worldofpeace commented on pkgs/development/python-modules/python-dotenv/default.nix:
```diff
@@ -0,0 +1,20 @@
+
+buildPythonPackage rec {
+  pname = "python-dotenv";
+  version = "0.9.1";
```
Newer version available https://github.com/theskumar/python-dotenv/releases/tag/v0.10.1

erikarvstedt replied:
Fixed

worldofpeace commented on pkgs/development/python-modules/django-filter/default.nix:
```diff
@@ -0,0 +1,23 @@
+  checkPhase = "${python.interpreter} runtests.py";
+
+  meta = with lib; {
+    description = "Django-filter is a reusable Django application for allowing users to filter querysets dynamically.";
```
```suggestion
    description = "A reusable Django application for allowing users to filter querysets dynamically";
```

erikarvstedt replied:
Fixed

worldofpeace commented on pkgs/development/python-modules/django-filter/default.nix:
```diff
@@ -0,0 +1,23 @@
+
+buildPythonPackage rec {
+  pname = "django-filter";
+  version = "2.0.0";
```
Newer version available https://github.com/carltongibson/django-filter/releases/tag/2.1.0

erikarvstedt replied:
Fixed: https://github.com/NixOS/nixpkgs/commit/2e9ce4e3d30b0ea312abbd7d68556c8dfc0b9d57

worldofpeace commented on pkgs/development/python-modules/django-filter/default.nix:
```diff
@@ -0,0 +1,23 @@
+buildPythonPackage rec {
+  pname = "django-filter";
+  version = "2.0.0";
+
```
We need to use the `disabled` parameter to respect https://github.com/carltongibson/django-filter/blob/8e917de6c90d3d5512d71bae05dc463ecaebc88b/setup.py#L59

erikarvstedt replied:
Fixed

worldofpeace commented on pkgs/development/python-modules/django-crispy-forms/default.nix:
```diff
@@ -0,0 +1,27 @@
+
+  meta = with lib; {
+    description = "The best way to have DRY Django forms";
+    homepage = http://github.com/maraujop/django-crispy-forms;
```
```suggestion
    homepage = https://github.com/maraujop/django-crispy-forms;
```

erikarvstedt replied:
Fixed

worldofpeace commented on pkgs/development/python-modules/django-cors-headers/default.nix:
```diff
@@ -0,0 +1,30 @@
+buildPythonPackage rec {
+  pname = "django-cors-headers";
+  version = "2.4.0";
+
```
We need to disable this for python 2.7.

worldofpeace commented on pkgs/development/python-modules/faker/default.nix:
```diff
@@ -31,6 +31,8 @@ buildPythonPackage rec {
   postPatch = ''
     substituteInPlace setup.py --replace "pytest>=3.8.0,<3.9" "pytest"
+    # Remove pycache dirs that cause the tests to fail with Python 3.6
+    find tests -path '*/__pycache__*' -delete
```
Just a heads up, this was resolved in master.

erikarvstedt replied:
Fixed

worldofpeace commented on pkgs/applications/office/paperless/default.nix:
```diff
@@ -0,0 +1,160 @@
+
+    nativeBuildInputs = [ makeWrapper ];
+
+    phases = [ "unpackPhase" "patchPhase" "buildPhase" "checkPhase" ];
```
I don't think we want to be doing this anymore https://github.com/NixOS/nixpkgs/issues/28910

erikarvstedt replied:
Fixed: https://github.com/NixOS/nixpkgs/commit/0dd454109793cea5226c2d0bfd2f9cbd587273e6

worldofpeace commented on nixos/modules/services/misc/paperless.nix:
```diff
@@ -0,0 +1,194 @@
+    port = mkOption {
+      type = types.int;
+      default = 28981;
+      description = "Server port.";
```
```suggestion
      description = "Server port to listen on.";
```

just for consistency.

erikarvstedt replied:
Fixed

worldofpeace commented:
More review incoming a bit later.

erikarvstedt committed:
pythonPackages.filemagic: init at 1.6

erikarvstedt committed:
pythonPackages.langdetect: init at 1.0.7

erikarvstedt committed:
pythonPackages.python-dotenv: init at 0.10.1

erikarvstedt committed:
pythonPackages.pytest-env: init at 0.6.2

erikarvstedt committed:
pythonPackages.pdftotext: init at 2.1.1

erikarvstedt committed:
pythonPackages.inotify-simple: init at 1.1.8

erikarvstedt committed:
pythonPackages.djangoql: init at 0.12.6

erikarvstedt committed:
paperless: init at 2.7.0

erikarvstedt committed:
paperless service: init

erikarvstedt commented:
https://github.com/NixOS/nixpkgs/commit/2e62f22688f7ec6145f916ac3b877b6ebcc7dd01: Rebased to master.

erikarvstedt commented:
https://github.com/NixOS/nixpkgs/commit/7a9ab89566ecb1b8026c0e032fa1e66381990b41: Squashed fixup commits.

erikarvstedt commented on pkgs/development/python-modules/django-filter/default.nix:
```diff
@@ -0,0 +1,23 @@
+  meta = with lib; {
+    description = "Django-filter is a reusable Django application for allowing users to filter querysets dynamically.";
+    homepage = https://github.com/carltongibson/django-filter;
+    license = licenses.bsdOriginal;
```
Fixed

erikarvstedt commented:
https://github.com/NixOS/nixpkgs/commit/603e473bdc5b20abd97dc201282a2a6e3d66fc04: Squashed fixup commits. [Old history](https://github.com/NixOS/nixpkgs/commits/deb594aa5606813fd3413baccd1d42ba90302af0).

(Sorry, the force-pushes (c4209b0 → 603e473), (603e473 → deb594a) were unintended.)

erikarvstedt commented:
https://github.com/NixOS/nixpkgs/commit/78bb5bce2d503a9e0359689b1817f7b8216f86cf: Rebased to `master`, squashed fixup commits.

worldofpeace commented:
@dotlambda Do you feel this is suitable?

Don't think it's fair to @erikarvstedt  to leave this unmerged any longer.

erikarvstedt commented:
I can understand if the review and maintanence burden of this feature seems unreasonable to you. I'd be fine with moving it to [NUR](https://github.com/nix-community/NUR) then.

worldofpeace commented:
> I can understand if the review and maintanence burden of this feature seems unreasonable to you. I'd be fine with moving it to [NUR](https://github.com/nix-community/NUR) then.

Oh no, I think this is perfectly suitable to have in nixpkgs.

dotlambda commented on nixos/modules/services/misc/paperless.nix:
```diff
@@ -0,0 +1,194 @@
+        systemd.services.paperless-consumer = {
+          description = "Paperless document consumer";
+          serviceConfig = {
+            PermissionsStartOnly = true;
```
In the future, this should be replaced by tmpfiles.d: https://github.com/NixOS/nixpkgs/pull/56265. But let's leave that for a future PR.

erikarvstedt replied:
`tmpfiles.d` really simplifies the implementation. Fixed here: https://github.com/NixOS/nixpkgs/pull/44086/commits/b5391a0db58a73b10f85e3bd32d73943122e9982

dotlambda commented on pkgs/applications/office/paperless/default.nix:
```diff
@@ -0,0 +1,168 @@
+    };
+  };
+
+  # Use Python 3.6 because dateparser is broken with 3.7 (https://github.com/NixOS/nixpkgs/issues/52766)
```
No longer true.

dotlambda commented on pkgs/applications/office/paperless/default.nix:
```diff
@@ -0,0 +1,168 @@
+  django_2_0 = pyPkgs: pyPkgs.django_2_1.overrideDerivation (_: rec {
+    pname = "Django";
+    version = "2.0.12";
+    name = "${pname}-${version}";
```
```suggestion
```

erikarvstedt replied:
For `overrideDerivation` , we have to update `name` manually.

dotlambda commented on pkgs/applications/office/paperless/withConfig.nix:
```diff
@@ -0,0 +1,68 @@
+    set -e
+    ${setupEnv}
+    ${extraCmds}
+    exec python $paperlessSrc/manage.py "$@"
```
```suggestion
    exec ${paperless.python.interpreter} $paperlessSrc/manage.py "$@"
```

erikarvstedt replied:
The correct form would be `paperless.pythonEnv.interpreter`.
`setupEnv` already handles all these details, so I'd rather not repeat them to keep it DRY.

dotlambda reviewed (approved):
Only some really minor stuff. And yes, this should definitely be merged.

erikarvstedt commented:
https://github.com/NixOS/nixpkgs/commit/b5651b3544efab70923b5e2759d158896cc2cf5e: Rebased to master
- Updated djangoql to the latest release
- Provide django-filter, django-crispy-forms as [private implementations](https://github.com/NixOS/nixpkgs/tree/69fd4abc27bb59f8d92dd65e69afe04395025c6a/pkgs/applications/office/paperless/python-modules), because 
  we use unreleased versions and patches to make them compatible with the current nixpkgs
  python package env.

worldofpeace commented on nixos/modules/services/misc/paperless.nix:
```diff
@@ -18,17 +18,13 @@ let
-      default = false;
-      description = ''
-        Enable Paperless.
+    enable = mkEnableOption ''
```
`mkEnableOption` does

https://github.com/NixOS/nixpkgs/blob/14ee8fefd7e5cb11d3e429e48a8f1c71f7ff9be9/lib/options.nix#L68

So it's not really right to use it for this.

erikarvstedt replied:
Is [this](https://github.com/NixOS/nixpkgs/pull/44086/commits/d0ed820adaeb6a022f3f746b999e8e8f5ebde3b1) okay, or too tricky?

worldofpeace replied:
Well `mkEnableOption` really just wants to take a `name` and there's a lot more text here that isn't part of the name.

erikarvstedt replied:
Alright, ditched it.

worldofpeace commented:
@erikarvstedt If you can rebase/squash this I can merge this.

worldofpeace commented:
Thank you @erikarvstedt for your patience (also very organized summary of changes which helps with reviews) and of course your contributions :sparkles: 

Likewise thanks to our reviewers :+1:

FRidh commented:
The attribute name is the attribute in the package set, and not the `pname` or `name` attribute in the expression.

erikarvstedt commented:
Huge thanks for your help and feedback, @dotlambda, @worldofpeace!

@FRidh, you meant to post that in another thread, right?
